system/lib/pthread/proxying.c

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#include <assert.h>
+#include <emscripten/html5.h>
+#include <emscripten/proxying.h>
+#include <emscripten/threading.h>
+#include <pthread.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define TASK_QUEUE_INITIAL_CAPACITY 128
+
+// TODO: Update this to take a `em_proxying_queue` argument.
+extern int _emscripten_notify_thread_queue(pthread_t target_thread,
+                                           pthread_t main_thread);
+
+typedef struct task {
+  void (*func)(void*);
+  void* arg;
+} task;
+
+// A task queue for a particular thread. Organized into a linked list of
+// task_queues for different threads.
+typedef struct task_queue {
+  // The target thread for this task_queue.
+  pthread_t thread;
+  // Recursion guard.
+  int processing;
+  // Ring buffer of tasks of size `capacity`. New tasks are enqueued at
+  // `tail` and dequeued at `head`.
+  task* tasks;
+  int capacity;
+  int head;
+  int tail;
+} task_queue;
+
+static int task_queue_empty(task_queue* tasks) {
+  return tasks->head == tasks->tail;
+}
+
+static int task_queue_full(task_queue* tasks) {
+  return tasks->head == (tasks->tail + 1) % tasks->capacity;
+}
+
+// Returns 1 on success and 0 on failure.
+static int task_queue_enqueue(task_queue* tasks, task t) {
+  if (task_queue_full(tasks)) {
+    // Allocate a larger task queue.
+    int new_capacity = tasks->capacity * 2;
+    task* new_tasks = malloc(sizeof(task) * new_capacity);
```

Comment:
Yes, but only for the case where the queue doesn't wrap around the end of the buffer. Since we need to malloc + memcpy + free on one code path anyway, it's simpler to just do it for both paths and share all the code except for the memcpy.